core/src/hlist.rs

```diff
@@ -794,75 +805,47 @@ where
     }
 }
 
-/// An alternative to AsRef that lets us create implementations for our recursive
-/// traits that takes references, without having to deal with weird impl overflows
+/// An alternative to AsRef that does not force the reference type to be a pointer itself.
+///
+/// This lets us create implementations for our recursive traits that take the resulting
+/// Output reference type, without having to deal with strange, spurious overflows
+/// that sometimes occur when trying to implement a trait for &'a T (see this comment:
+/// https://github.com/lloydmeta/frunk/pull/106#issuecomment-377927198)
 ///
 /// This is essentially From, but the more specific nature of it means it's more ergonomic
-/// in actual usage
-pub trait ToRef {
+/// in actual usage.
+pub trait ToRef<'a> {
     type Output;
-    fn to_ref(self) -> Self::Output;
+
+    #[inline(always)]
+    fn to_ref(&'a self) -> Self::Output;
 }
 
-impl <'a> ToRef for &'a HNil {
-    type Output = &'a HNil;
-    fn to_ref(self) -> Self::Output {
-        &HNil
-    }
+impl<'a> ToRef<'a> for HNil {
+    type Output = HNil;
 
+    #[inline(always)]
+    fn to_ref(&'a self) -> Self::Output {
+        HNil
+    }
 }
 
-impl <'a, H, Tail> ToRef for &'a HCons<H, Tail>
-where &'a Tail: ToRef {
-    type Output = HCons<&'a H, <&'a Tail as ToRef>::Output>;
-    fn to_ref(self) -> Self::Output {
+impl<'a, H, Tail> ToRef<'a> for HCons<H, Tail>
+    where
+        H: 'a,
+        Tail: ToRef<'a>,
```

Comment:
I can tell you use clion/intellij from how these where bounds are getting indented =P

Reply:
Fixed! Oddly enough, I `cargo +nightly fmt`ed and copy and pasted the formatted new stuff, but IntelliJ was insistent on screwing it up.

Reply:
Yep, can't trust copy and paste in intellij.  I often find myself fixing the same where bounds multiple times thanks to this. =P